proto/rill/admin/v1/api.proto

```diff
@@ -555,3 +569,9 @@ message UserInvite {
   string role = 2;
   string invited_by = 3;
 }
+
+
+message PageToken {
+  // cursor is kept as an array since it may not be possible to use one field to unqiuely identify in all cases
+  repeated string cursor = 1;
+}
```

Comment:
1. Since this isn't part of the external API, move to a file `internal.proto` in `proto/rill/admin/v1`
2. Don't need to take a generic approach to the fields – empty fields are omitted in protobuf serialization, so it's a needless optimization. Better to have explicit fields like `string string_value = 1`, and can add other fields if needed in the future (like `int64 int_value = 2`, `repeated string string_values = 3`, etc.).

Reply:
2. It will be very weird to see a `PageToken` proto in future with both `string_value` and `string_values`. 
```
message PageToken {
  string string_value = 1;
  repeated string string_values = 2;
}
```
This will just add tons of confusion. 

The necessity of having multiple columns as cursor isn't very far fetched either. Imagine getting all projects for users across all their organisations.
Also I find keeping different datatype fields confusing as well. Its neither generic nor specific.
I would either choose an API specific token or a generic token with repeated fields(even if that meant I have to cast in the code).

Reply:
I agree that some cases might warrant an API specific format.

But if going with a generic format, it's better to at least use the type system as much as possible. E.g. if you need an int offset, coercing it to a string array is weird. Not sure what the right solution is, some other ideas:

- Use a `oneof` type, for example: https://github.com/protocolbuffers/protobuf/blob/0ae328929ee2b6378d8cb3bb2c26700922098910/src/google/protobuf/struct.proto#L62
- Have specific messages for different types, like:
```proto
message StringCursor {
  string value = 1;
}

// Etc.
```
- Have something like `repeated google.protobuf.Value`, but you'd need a custom type since `google.protobuf.Value` can't contain ints.
- Use semantic field names (note this is exactly what you would implicitly do if serializing a map):
```proto
message PageToken {
  oneof property {
    string id = 1;
    string name = 2;
    int32 offset = 3;
  }
}
```

Reply:
Also to add its not uncommon to have an array of string for such use cases. Take an example of [grpc-metdata](https://github.com/grpc/grpc-go/blob/master/Documentation/grpc-metadata.md#constructing-metadata) the key is `string` and the value is a `[]string`. 

I would not add `offset` in the cursor but would instead add two new fields for `offset` and `page`. I would expect to set `cursor` when using cursor based pagination or `offset` and `page` when using offset based pagination. Alternatively I would create another type of PageToken and would probably use that. The goal with cursor as `[]string` is not to create a generic token for all use cases but to be able to adapt for different cursors that's why the field name cursor.

1. `oneof` can't be repeated and we may need a string and int cursor both at the same time. I would need to wrap this in another message and it will be same as the third idea in the list.
2. I like this idea. Its simple and self explanatory that the token is a `stringcursor`.
3. This one works but feel its complicated for the use case. It makes sense to use this if I am not marshalling the data as is the case with `json`. Here I clearly know what `type` I will be getting. 
4. Sorry I didn't really understand this.

Reply:
I think we're a little confused on terminology here – a "cursor" can be anything that marks the next place to fetch from, can also be an "offset". 

We would do best to avoid using both "cursor" and "page token" since they are basically the same. I think introducing a `StringPageToken` containing one `string value = 1` field makes sense for now. In the future, we can then introduce `StringsPageToken`, `IntPageToken`, `APISpecificPageToken`, etc. types as needed.

How does that sound?

Reply:
Fundamentally agree but have always found `offset` to be considered different from `cursor`.

Yeah `StringPageToken ` sounds good and we can add more types if required. Will change it to `StringPageToken`.